### PR TITLE
HAWQ-927. Pass ProjectionInfo data to PXF

### PR DESCRIPTION
**Pull request by kavinderd**
This commit makes the necessary modifications to the HAWQ side of
the codebase to add a list of indices of projected columns

### CODE REVIEW

File: src/backend/access/external/pxfheaders.c [hunk 1]
```diff
@@ -158,6 +165,29 @@ static void add_tuple_desc_httpheader(CHURL_HEADERS headers, Relation rel)
 	pfree(formatter.data);
 }
 
+static void add_projection_desc_httpheader(CHURL_HEADERS headers, ProjectionInfo *projInfo) {
+	int i;
+	char long_number[32];
```
Comment (sansanichfb) [inline src/backend/access/external/pxfheaders.c]: Why 32?

File: src/include/access/fileam.h [hunk 2]
```diff
@@ -63,6 +63,17 @@ typedef struct ExternalInsertDescData
 
 typedef ExternalInsertDescData *ExternalInsertDesc;
 
+/*
+ * ExternalSelectDescData is used for storing state related
+ * to selecting data from an external table.
+ */
+typedef struct ExternalSelectDescData
```
Comment (sansanichfb) [inline src/include/access/fileam.h]: Why do we introduce new struct if we can just use ProjectionInfo?

Comment (kavinderd) [inline src/include/access/fileam.h]: That's an option but since we plan on passing other data like the type of query, etc to PXF I think it will be more maintainable to have this encapsulating struct so that we don't need to keep changing the method signatures.

What are your thoughts?

File: src/backend/access/external/pxfheaders.c (hunk 1 above)
Comment (kavinderd) [inline src/backend/access/external/pxfheaders.c]: Since we are converting `list_length(projInfo->pi_targetlist)`, we need an array thats 32 chars long to match the bit length of `int`.

Comment (sansanichfb) [inline src/backend/access/external/pxfheaders.c]: You can try something like:
char long_number[(sizeof int) * 8];

File: src/include/access/fileam.h (hunk 2 above)
Comment (sansanichfb) [inline src/include/access/fileam.h]: Makes sense

File: src/include/access/url.h [hunk 3]
```diff
@@ -154,12 +154,15 @@ typedef struct extvar_t
  	char GP_CSVOPT[13]; /* "m.x...q...h." former -q, -h and -x options for gpfdist.*/
 } extvar_t;
 
+//struct ExternalSelectDescData;
```
Comment (sansanichfb) [inline src/include/access/url.h]: Cleanup comments?

**Comment by kavinderd** [comment]
Please review @xunzhang @paul-guo- @huor

**Comment by shivzone** [comment]
I don't think u need more than 10 bytes for long_number to store a long int.

File: src/backend/access/external/fileam.c [hunk 4]
```diff
@@ -454,6 +454,21 @@ external_stopscan(FileScanDesc scan)
...
+/*	----------------
+ *		external_getnext_init - prepare ExternalSelectDesc struct before external_getnext
+/*	----------------
+ */
+
+ExternalSelectDesc
+external_getnext_init(PlanState *state) {
+	ExternalSelectDesc desc = (ExternalSelectDesc) palloc0(sizeof(ExternalSelectDescData));
+	if (state != NULL && state->ps_ProjInfo != NULL) {
+		desc->projInfo = state->ps_ProjInfo;
```
Comment (shivzone) [inline src/backend/access/external/fileam.c]: Second check is redundant

File: src/backend/access/external/fileam.c [hunk 5]
```diff
@@ -462,7 +462,7 @@ external_stopscan(FileScanDesc scan)
 ExternalSelectDesc
 external_getnext_init(PlanState *state) {
 	ExternalSelectDesc desc = (ExternalSelectDesc) palloc0(sizeof(ExternalSelectDescData));
-	if (state != NULL && state->ps_ProjInfo != NULL) {
+	if (state != NULL) {
 		desc->projInfo = state->ps_ProjInfo;
 	} else {
```
Comment (xunzhang) [inline src/backend/access/external/fileam.c]: You should also delete this `else-branch`, it is unnecessary too.

File: src/backend/access/external/pxfheaders.c [hunk 6]
```diff
@@ -158,6 +166,29 @@ static void add_tuple_desc_httpheader(CHURL_HEADERS headers, Relation rel)
...
+static void add_projection_desc_httpheader(CHURL_HEADERS headers, ProjectionInfo *projInfo) {
+	int i;
+	char long_number[sizeof(int32) * 8];
+	int *varNumbers = projInfo->pi_varNumbers;
+	StringInfoData formatter;
+	initStringInfo(&formatter);
+
+    /* Convert the number of projection columns to a string */
+    pg_ltoa(list_length(projInfo->pi_targetlist), long_number);
+    churl_headers_append(headers, "X-GP-ATTRS-PROJ", long_number);
```
Comment (xunzhang) [inline src/backend/access/external/pxfheaders.c]: indent~

File: src/backend/access/external/pxfheaders.c [hunk 7]
```diff
@@ -123,7 +130,8 @@ static void add_alignment_size_httpheader(CHURL_HEADERS headers)
  */
 static void add_tuple_desc_httpheader(CHURL_HEADERS headers, Relation rel)
 {	
-    char long_number[32];	
+    char long_number[sizeof(int32) * 8];
```
Comment (xunzhang) [inline src/backend/access/external/pxfheaders.c]: I like this modification!

File: src/backend/access/external/fileam.c [hunk 8]
```diff
@@ -454,6 +454,21 @@ external_stopscan(FileScanDesc scan)
...
 }
 
+/*	----------------
+ *		external_getnext_init - prepare ExternalSelectDesc struct before external_getnext
+/*	----------------
+ */
+
+ExternalSelectDesc
+external_getnext_init(PlanState *state) {
+	ExternalSelectDesc desc = (ExternalSelectDesc) palloc0(sizeof(ExternalSelectDescData));
```
Comment (xunzhang) [inline src/backend/access/external/fileam.c]: I just want to make sure when does `desc` object be freed?

File: src/backend/access/external/fileam.c [hunk 9]
```diff
@@ -454,6 +454,21 @@ external_stopscan(FileScanDesc scan)
 	}
 }
 
+/*	----------------
+ *		external_getnext_init - prepare ExternalSelectDesc struct before external_getnext
+/*	----------------
```
Comment (xunzhang) [inline src/backend/access/external/fileam.c]: remove `/` here.

File: src/backend/access/external/pxfheaders.c [hunk 10]
```diff
@@ -158,6 +166,29 @@ static void add_tuple_desc_httpheader(CHURL_HEADERS headers, Relation rel)
...
+	char long_number[sizeof(int32) * 8];
+	int *varNumbers = projInfo->pi_varNumbers;
+	StringInfoData formatter;
+	initStringInfo(&formatter);
+
+    /* Convert the number of projection columns to a string */
+    pg_ltoa(list_length(projInfo->pi_targetlist), long_number);
+    churl_headers_append(headers, "X-GP-ATTRS-PROJ", long_number);
+
+	for(i = 0; i < list_length(projInfo->pi_targetlist); i++) {
```
Comment (xunzhang) [inline src/backend/access/external/pxfheaders.c]: @kavinderd What' the code logic of PXF end to parse this header? Does it be included in another pull request?

File: src/backend/access/external/fileam.c (hunk 8 above)
Comment (kavinderd) [inline src/backend/access/external/fileam.c]: I missed adding `pfree` for `desc`. I added it to the end of `ExternalNext()`

File: src/backend/access/external/pxfheaders.c (hunk 10 above)
Comment (kavinderd) [inline src/backend/access/external/pxfheaders.c]: Yes it will be in another PR related to [this](https://issues.apache.org/jira/browse/HAWQ-583?jql=project%20%3D%20HAWQ%20AND%20resolution%20%3D%20Unresolved%20AND%20assignee%20%3D%20kavinderd%20ORDER%20BY%20priority%20DESC) Jira.

**Comment by xunzhang** [comment]
LGTM. Remember rebasing the commit before checking in. 🍺

**Comment by shivzone** [comment]
+1
